Use generic error handling for 4XX empty responses

adev-code opened the pull request:
### Problem

When AWS services return 4XX errors with empty bodies, users receive XML parsing errors instead of meaningful HTTP status messages. For example, service returns HTTP 413 with an empty body when request size limits are exceeded, but users see:
```
botocore.parsers.ResponseParserError: Unable to parse response (no element found: line 1, column 0), invalid XML received.
```
### Solution

This PR extends generic error handling from 5XX to 4XX status codes, preventing parsing failures and surfacing the actual HTTP error, for example:
```
botocore.exceptions.ClientError: An error occurred (413) when calling the PutDashboard operation: Content Too Large
```
### Backwards Compatibility

- Preserves all existing 5XX behavior
- Only affects 4XX responses with empty/HTML bodies that would previously cause parsing failures

adev-code committed:
Add HTTP status context to XML parsing errors using PEP 678 notes

When XML parsing fails due to empty or malformed responses, users now
receive the actual HTTP error response from the service alongside the
parsing error, providing useful information about what actually happened.

Before:
  ResponseParserError: Unable to parse response (no element found:
  line 1, column 0), invalid XML received. Further retries may succeed: b''

After:
  ResponseParserError: Unable to parse response (no element found:
  line 1, column 0), invalid XML received. Further retries may succeed: b''
  HTTP 413: Content Too Large

This exposes the real service error (HTTP 413: Content Too Large) that
was previously hidden behind cryptic XML parsing failures, giving users
actionable information about why their request failed.

adev-code committed:
Fix for Fix test assertions to work across Python ver. 3.11 & 3.12

The HTTP 413 status message differs between Python versions:
- Python 3.13: "Content Too Large"
- Python 3.11: "Request Entity Too Large"

Updated test assertions to use assertIn("HTTP 413:") instead of exact
string matching to ensure tests pass across supported Python versions
while still verifying that HTTP status context is properly added to
XML parsing errors.

adev-code committed:
Add Python 3.9-3.10 compatibility for PEP 678 exception notes

Add hasattr(e, 'add_note') checks before calling add_note() to ensure
compatibility with Python 3.9 and 3.10 where PEP 678 exception notes
are not available. The enhanced HTTP status context will only be added
in Python 3.11+ while maintaining backward compatibility.

nateprewitt commented on botocore/parsers.py:
```diff
@@ -598,7 +598,7 @@ def _do_error_parse(self, response, shape):
                 response['status_code'], ''
             )
-            if status_message:
+            if status_message and hasattr(e, 'add_note'):
```
I think this requires Python 3.11+ so this is probably not the approach we want to take to make sure this works everywhere.

adev-code committed:
Add HTTP status context to XML parsing error messages

Enhanced ResponseParserError messages to include HTTP status codes and
messages when XML parsing fails, providing better debugging context for
users while maintaining full backward compatibility.

Changes:
- Modified QueryParser._do_error_parse() to append HTTP status context
- Modified RestXMLParser._parse_error_from_body() to append HTTP status context
- Enhanced error messages format: "original error (HTTP 413: Content Too Large)"
- Preserves original ResponseParserError exception type for compatibility
- Added comprehensive tests for both parser classes
- Tests handle Python version differences in HTTP status messages

adev-code committed:
Refactor QueryParser error handling for consistency and clarity

- Improve code readability while maintaining identical functionality

adev-code committed:
Delete .python-version

adev-code committed:
Use generic error handling for 4XX empty responses

Extend generic error response detection from 5XX to 4XX status codes
to prevent XML parsing failures when services return empty bodies.

Fixes parsing crashes for cases like AWS Services returning
HTTP 413 with empty body, now shows "Content Too Large" instead of
"Unable to parse response (no element found: line 1, column 0)".

adev-code committed:
Remove outdated tests and fix

adev-code committed:
nit, remove my comment on test parsers

robert-schmidtke commented:
Hi, I believe I am running into this issue and was wondering whether this fix was still being pursued in some way.
